akka-stream/src/main/scala/akka/stream/impl/fusing/GraphStages.scala

```diff
@@ -319,15 +320,11 @@ import scala.concurrent.{ Future, Promise }
 
           override def onDownstreamFinish(): Unit = {
             if (!materialized.isCompleted) {
-              // make sure we always yield the matval if possible, even if downstream cancelled
-              // before the source was materialized
-              val matValFuture = futureSource.map { gr ⇒
-                // downstream finish means it cancelled, so we push that signal through into the future materialized source
-                Source.fromGraph(gr).to(Sink.cancelled)
-                  .withAttributes(attr)
-                  .run()(subFusingMaterializer)
-              }(ExecutionContexts.sameThreadExecutionContext)
-              materialized.completeWith(matValFuture)
+              // we used to try to materialize the "inner" source here just to get
+              // the materialized value, but that is not safe and may cause the graph shell
+              // to leak/stay alive after the stage completes
+
+              materialized.tryFailure(new StreamDetachedException("Stream finished before future source completed"))
```

Comment:
I think it makes sense, but I was thinking we should ask the original author and that is you ;-)
https://github.com/akka/akka/pull/22360/files#diff-a5b5083cbae1b01b3fc1ba21536091a6R307

Reply:
I did discuss this with myself before changing it and I had no objections to this change.

Reply:
More specific wording perhaps? "Stream cancelled before Source Future completed"?

Semantics sound good though